examples/spring-hypermedia/src/main/java/io/wcm/caravan/rhyme/spring/impl/SpringRhymeImpl.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * #%L
+ * wcm.io
+ * %%
+ * Copyright (C) 2021 wcm.io
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package io.wcm.caravan.rhyme.spring.impl;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.CacheControl;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.ResponseEntity.BodyBuilder;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.annotation.RequestScope;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.wcm.caravan.rhyme.api.Rhyme;
+import io.wcm.caravan.rhyme.api.RhymeBuilder;
+import io.wcm.caravan.rhyme.api.common.HalResponse;
+import io.wcm.caravan.rhyme.api.resources.LinkableResource;
+import io.wcm.caravan.rhyme.api.spi.HalResourceLoader;
+import io.wcm.caravan.rhyme.spring.api.SpringRhyme;
+import io.wcm.caravan.rhyme.spring.api.UrlFingerprinting;
+
+/**
+ * Implementation of the request-scoped {@link SpringRhyme} interface that creates a single {@link Rhyme} instance
+ * using a caching {@link WebClientHalResourceLoader}, {@link SpringExceptionStatusAndLoggingStrategy} and the
+ * {@link SpringRhymeDocsIntegration}. It also contains methods to render a {@link LinkableResource} (or any exception),
+ * as a Spring {@link ResponseEntity}, but these methods are not made public because these conversions are handled by
+ * the {@link LinkableResourceMessageConverter} and {@link VndErrorHandlingControllerAdvice}.
+ */
+@Component
+@RequestScope
+class SpringRhymeImpl implements SpringRhyme {
+
+  private static final Logger log = LoggerFactory.getLogger(SpringRhymeImpl.class);
+
+  private static final SpringExceptionStatusAndLoggingStrategy EXCEPTION_STRATEGY = new SpringExceptionStatusAndLoggingStrategy();
+
+  private final HttpServletRequest request;
+
+  private final Rhyme rhyme;
+
+  private ResponseEntity<JsonNode> renderedResponse;
+
+  SpringRhymeImpl(@Autowired HttpServletRequest httpRequest,
+      @Autowired HalResourceLoader resourceLoader,
```

Comment:
This auto-wiring of the HalResourceLoader currently works well only because there are only two possible implementations:

When the application is started regularly, it will use the WebClientHalResourceLoader, as it's the only @Component implementing that interface on the class path.

When the integration tests are being executed, the MockMvcHalResourceLoader is used instead, because it's on the test classpath and annotated with `@Primary` annotation. 

This is exactly what we want for this example, but if the SpringRhymeImpl code is moved into a common integration module, there needs to be a way for developers to control which HalResourceLoader is used. As soon as you require HTTP authentication or other configuration of the HTTP client (or caching), you would currently need to build your own HalResourceLoader implementation and make this SpringRhymeImpl instance pick it up somehow

Any idea what's the easiest way to achieve this?

Reply:
I've improved this a bit in c833778e209dbdfd2e149f9d1e24c1d0bb7c2c14 by no longer having a MockMvcHalResourceLoader that is a `@Primary @Component'.

Instead, each integration test that wants to load such a loader must add an`@Import(MockMvcHalResourceLoaderConfiguration.class)`instead.

At least this special HalResourceLoader is no longer picked up automatically (completely replacing the default implementation) just because it's in the test classpath.